code/modules/power/cell.dm

```diff
@@ -383,3 +359,136 @@
+
+/obj/item/stock_parts/cell/pulse/pistol //10 pulse shots
+	name = "pulse pistol power cell"
+	maxcharge = 2000
+
+//	---------------------
+//		GUN CELLS
+//	---------------------
+
+/obj/item/stock_parts/cell/gun
+	name = "power supply"
+	desc = "a reinforced cell capable of releaseing short, high intensity electricity bursts without exploding."
+	icon = 'icons/obj/power.dmi'
+	icon_state = "g-cell"
+	maxcharge = 1500
+	custom_materials = list(/datum/material/glass=60)
+	chargerate = 150
+	var/wired = FALSE
+
+/obj/item/stock_parts/cell/gun/update_icon()
+	cut_overlays()
+	if(grown_battery)
+		add_overlay(image('icons/obj/power.dmi',"grown_wires"))
+	if(charge < 0.01)
+		return
+	else if(charge/maxcharge >=0.995)
+		add_overlay("g-cell-o2")
+	else
+		add_overlay("g-cell-o1")
+
+/obj/item/stock_parts/cell/gun/empty/Initialize()
+	. = ..()
+	charge = 0
+	update_icon()
+
+/obj/item/stock_parts/cell/gun/update_icon()
+	cut_overlays()
+	if(grown_battery)
+		. += mutable_appearance('icons/obj/power.dmi', "grown_wires")
+	if(charge < 0.01)
+		return
+	else if(charge/maxcharge >=0.995)
+		add_overlay("g-cell-o2")
+	else
+		add_overlay("g-cell-o1")
+
+/obj/item/stock_parts/cell/gun/rechargeable
+	self_recharge = 100
+
+/obj/item/stock_parts/cell/gun/super
+	name = "improved power supply"
+	maxcharge = 2000
+	custom_materials = list(/datum/material/glass=300)
+	chargerate = 175
+
+/obj/item/stock_parts/cell/gun/super/empty/Initialize()
+	. = ..()
+	charge = 0
+	update_icon()
+
+/obj/item/stock_parts/cell/gun/hicap
+	name = "high capacity power supply"
+	icon_state = "ug-cell"
+	maxcharge = 3000
+	custom_materials = list(/datum/material/glass=300)
+	chargerate = 100
+
+/obj/item/stock_parts/cell/gun/hicap/empty/Initialize()
+	. = ..()
+	charge = 0
+	update_icon()
+
+/obj/item/stock_parts/cell/gun/frec
+	name = "fast recharge power supply"
+	icon_state = "ug-cell"
+	maxcharge = 1500
+	custom_materials = list(/datum/material/glass=300)
+	chargerate = 400
+
+/obj/item/stock_parts/cell/gun/frec/empty/Initialize()
+	. = ..()
+	charge = 0
+	update_icon()
+
+/obj/item/stock_parts/cell/gun/mini
+	name = "compact power supply"
+	icon_state = "mg-cell"
+	custom_materials = list(/datum/material/glass=300)
+	maxcharge = 600
+	chargerate = 100
+
+/obj/item/stock_parts/cell/gun/mini/empty/Initialize()
+	. = ..()
+	charge = 0
+	update_icon()
+
+/obj/item/stock_parts/cell/gun/mini/update_icon()
+	cut_overlays()
+	if(grown_battery)
+		. += mutable_appearance('icons/obj/power.dmi', "grown_wires")
+	if(charge < 0.01)
+		return
+	else if(charge/maxcharge >=0.995)
+		add_overlay("mg-cell-o2")
+	else
+		add_overlay("mg-cell-o1")
+
+
+/obj/item/stock_parts/cell/gun/atech
+	name = "alien power supply"
+	maxcharge = 1000
+	custom_materials = list(/datum/material/glass=60)
+	self_recharge = 100
+
+/obj/item/stock_parts/cell/gun/mini/atech
+	name = "compact alien power supply"
+	custom_materials = list(/datum/material/glass=300)
+	maxcharge = 300
+	self_recharge = 100
```

Comment:
`self_recharge` is a *flag* used to determine whether cells should charge as much as `chargerate` every 2 seconds at `/obj/item/stock_parts/cell/process()`, which every cell automatically gets to run at least once since there's `START_PROCESSING` at `/obj/item/stock_parts/cell/Initialize`. You have to take a look at it.

Reply:
Sure. Not sure if self charging alien tech guncells is even a good idea, though.

Reply:
its not